src/_P092_DLbus.ino

```diff
@@ -0,0 +1,1288 @@
+   UVR31, UVR1611 and UVR 61-3 devices.
+
+   The selected input needs a voltage divider as follows because the DL-Bus runs on 12 volts for
+	 following devices: UVR31, UVR42, UVR64, HZR65, EEG30 and TFM66
+   DLbus@12V - 8k6 - input@3.3V - 3k3 - ground
+
+   For following devices just a pull up resistor is needed if the device is used stand alone:
+	 UVR1611, UVR61-3 and ESR21
+
+	 @uwekaditz 2019-12-15 Memory usage optimized
+	 CHG: Moved the array for the received bit changes to stativ uint_8t, the ISR call uses only a volatile pointer to it
+	 CHG: some more defines and name changes for better explanation
+
+	 @uwekaditz 2019-12-14 Timing optimized
+	 CHG: Removed the while (P092_receiving) loop.
+	 CHG: Starting of the receiving and processing of the received bit stream are now done in the PLUGIN_ONCE_A_SECOND call
+			  PLUGIN_READ call just uses the already processed data
+
+	 @uwekaditz 2019-12-08 Inital commit to mega
+
+\**************************************************/
+
+#define PLUGIN_092
+#define PLUGIN_ID_092         92
+//#define PLUGIN_092_DEBUG    // additional debug messages in the log
+#define PLUGIN_NAME_092       "Heating - DL-Bus (Technische Alternative)"
+#define PLUGIN_VALUENAME1_092 "Value"
+
+#define P092_DLbus_OptionCount 8
+#define P092_DLbus_ValueCount 1
+#define P092_DLbus_DeviceCount 5
+
+void P092_Pin_changed() ICACHE_RAM_ATTR;
+volatile uint8_t P092_DLB_Pin = 0xFF;
+volatile boolean P092_receiving = false;            // receiving flag
+volatile boolean P092_AllBitsReceived = false;
+uint8_t P092_Last_DLB_Pin;
+boolean P092_init = false;
+boolean P092_ReceivedOK = false;
+uint32_t P092_LastReceived = 0;
+uint8_t P092_MaxIdx[P092_DLbus_OptionCount];
+int P092_ValueType[P092_DLbus_ValueCount];
+int P092_ValueIdx[P092_DLbus_ValueCount];
+struct _P092_DataStruct
+{
+  uint8_t				DataBytes;
+  uint8_t       DeviceByte0;
+  uint8_t       DeviceByte1;
+  uint8_t       DeviceBytes;
+  uint8_t       DontCareBytes;
+  uint8_t       TimeStampBytes;
+  uint8_t       MaxSensors;
+  uint8_t       MaxExtSensors;
+  uint8_t       OutputBytes;
+  uint8_t       SpeedBytes;
+  uint8_t       AnalogBytes;
+  uint8_t       VolumeBytes;
+  uint8_t       MaxHeatMeters;
+  uint8_t       CurrentHmBytes;
+  uint8_t       MWhBytes;
+
+  uint16_t      DLbus_MinPulseWidth;
+  uint16_t      DLbus_MaxPulseWidth;
+  uint16_t      DLbus_MinDoublePulseWidth;
+  uint16_t      DLbus_MaxDoublePulseWidth;
+  uint8_t       IdxSensor;
+  uint8_t       IdxExtSensor;
+  uint8_t       IdxOutput;
+  uint8_t       IdxDrehzahl;
+  uint8_t       IdxAnalog;
+  uint8_t       IdxHmRegister;
+  uint8_t       IdxVolume;
+  uint8_t       IdxHeatMeter1;
+  uint8_t       IdxkWh1;
+  uint8_t       IdxMWh1;
+  uint8_t       IdxHeatMeter2;
+  uint8_t       IdxkWh2;
+  uint8_t       IdxMWh2;
+  uint8_t       IdxHeatMeter3;
+  uint8_t       IdxkWh3;
+  uint8_t       IdxMWh3;
+  uint8_t       IdxCRC;
+} P092_DataSettings;
```

Comment:
I understand your concern about the limited ressources on the ESP82 devices.
Nevertheless I need to explain, why `PluginTaskData_base ` is not working for this plug_in.
1. Only one controller device type can be  connected to a DL bus instance (ESR21, UVR31, UVR1611 or UVR 61-3).
2.  The data read and decoded from the DL bus are global for all task using the P092_DLbus-plug-in because they are read only by the global interrupt function of the IO pin (CONFIG_PIN1). The decoding happens also only once after all bytes are received.
3. A new device task must be created for each sensor value to be extracted from the DL bus (using the same device type and IO pin!). This new task only extracts the data for the requested sensor value from the received data (golabl, see 2.). I had to program this restriction because up to 37 sensor values can be combined in one DL bus frame (depending on the controller device type) and the ValueCount of any device task is limited to 4 (if I'm not wrong).

It would be very helpful for me if you have some example
- how to use the same pointer to allocated memory from different device tasks of the same type and
- how to allocate the memory to this pointer once the first device tasks is created and
- how to free the memory if the last device tasks of the same type is deleted.

I need to test the other changes I already made on my real system before I can upload the new source code.

Reply:
You can also create a single struct (e.g. `p092_datastruct`) for this shared data and then just define a "global" pointer to this (initialized to `null_ptr`).
As soon as you init the first task for this, you will create a `new p092_datastruct;` and assign it to this pointer.
As long as the pointer is still a `null_ptr`, then the task is not initialized (and also not taking memory).
If you really need it, you will use resources and that's fine then.

Reply:
See my last changes, I put as much as possible in a class, only created while using the P092 plug in.